internal/common/expand_into_runfiles.bzl

```diff
@@ -15,47 +15,91 @@
     paths = ctx.expand_location(input, targets)
-    return " ".join([_short_path_to_runfiles_manifest_path(ctx, p, targets) for p in paths.split(" ")])
+    return " ".join([_rootpath_to_runfiles_manifest_path(ctx, p, targets) for p in paths.split(" ")])
 
-# Convert a short_path in the execroot to the runfiles manifest path
-def _short_path_to_runfiles_manifest_path(ctx, path, targets):
+# Convert an runfiles rootpath to a runfiles manifestpath.
+# Runfiles rootpath is returned from ctx.expand_location $(rootpath) and $(rootpaths):
+# - ./file
+# - path/to/file
+# - ../external_repo/path/to/file
+# This is converted to the runfiles manifest path of:
+# - repo/path/to/file
+def _rootpath_to_runfiles_manifest_path(ctx, path, targets):
     if path.startswith("../"):
         return path[len("../"):]
     if path.startswith("./"):
         path = path[len("./"):]
-    elif path.startswith(ctx.bin_dir.path):
-        path = path[len(ctx.bin_dir.path + "/"):]
-    elif path.startswith(ctx.genfiles_dir.path):
-        path = path[len(ctx.genfiles_dir.path + "/"):]
     return ctx.workspace_name + "/" + path
 
-# Expand $(location) and $(locations) to runfiles short path
-def _expand_locations(ctx, input, targets):
-    paths = ctx.expand_location(input, targets)
-    return " ".join([_short_path_to_runfiles_short_path(ctx, p, targets) for p in paths.split(" ")])
-
-# Convert a short_path in the execroot to the runfiles short path
-def _short_path_to_runfiles_short_path(ctx, path, targets):
-    path = path.replace(ctx.bin_dir.path + "/external/", "../", 1)
-    path = path.replace(ctx.bin_dir.path + "/", "", 1)
-    path = path.replace(ctx.genfiles_dir.path + "/external/", "../", 1)
-    path = path.replace(ctx.genfiles_dir.path + "/", "", 1)
-    return path
-
 def expand_location_into_runfiles(ctx, input, targets = []):
-    """Expands all $(location ...) templates in the given string by replacing $(location //x) with the path
-    in runfiles of the output file of target //x. Expansion only works for labels that point to direct dependencies
+    """Expands all $(execpath ...), $(rootpath ...) and legacy $(location ...) templates in the
+    given string by replacing with the expanded path. Expansion only works for labels that point to direct dependencies
     of this rule or that are explicitly listed in the optional argument targets.
 
-    Path is returned in runfiles manifest path format such as `repo/path/to/file`. This differs from how $(location)
-    and $(locations) expansion behaves in expansion the `args` attribute of a *_binary or *_test which returns
-    the runfiles short path of the format `./path/to/file` for user repo and `../external_repo/path/to/file` for external
-    repositories. We may change this behavior in the future with $(mlocation) and $(mlocations) used to expand
-    to the runfiles manifest path.
+    See https://docs.bazel.build/versions/master/be/make-variables.html#predefined_label_variables.
```

Comment:
this is a nice and thorough docstring but does it appear anywhere after stardoc runs? will users ever see it? Is it a copy of the one on params_file? if so maybe just point to there so we don't end up with the copies diverging

Reply:
Its is slightly different from the param_files one and a subset of the nodejs_binary templated_args docstring. Not super thrilled with the duplication so I could point to either of those since params_file and nodejs_binary templated_args both use this helper function. OTOH, it is strange to document the helper function at the point it is used at and this docstring was intended more for rules_nodejs maintainers.

Reply:
Thinking on this a bit more, I'll make the helper function docstring briefer and point to the Bazel docs & to the nodejs_binary & params_file docstrings for more info.

Reply:
Done.